Tests/PackageGraphTests/PackageGraphTests.swift

```diff
@@ -1255,4 +1263,551 @@ class PackageGraphTests: XCTestCase {
         let store = try PinsStore(pinsFile: AbsolutePath("/pins"), fileSystem: fs)
         XCTAssertThrows(StringError("duplicated entry for package \"Yams\""), { try store.restore(from: json) })
     }
+
```

Comment:
@neonichu @abertelrud would love to get your review on the different permutations captured in these new tests 😅 and hear your opinion on:
1. the diagnostics make sense and are better than before, I think they are!
2. the backward compatibility 5.2 < 5.4 looks reasonable - we allow users to use dependency::name but no longer encourage it
3. do we need to capture more permutations?